src/main/java/com/bobocode/service/UrlShortenerService.java

```diff
@@ -0,0 +1,50 @@
+package com.bobocode.service;
+
+import com.bobocode.dto.UrlInfoRequest;
+import com.bobocode.entity.ShortUrl;
+import com.bobocode.exception.InvalidUrlException;
+import com.bobocode.exception.NoUrlFoundException;
+import com.bobocode.repository.UrlRepository;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.validator.routines.UrlValidator;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.retry.annotation.Retryable;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class UrlShortenerService {
+
+  private final UrlRepository urlRepository;
+
+  @Transactional
+  @Retryable
+  public String shorten(UrlInfoRequest urlInfo) {
+    checkUrlIsValid(urlInfo.url());
```

Comment:
@kebann generally speaking, it's not good idea to open a DB transaction and then perform input data validation.

Reply:
extracted this logic into a custom javax validator